src/audio/dsm/sof_dsm_inf.c

```diff
@@ -0,0 +1,272 @@
+	memset(hsof_dsm->buf.stage_fb, 0, DSM_FB_BUF_DB_SZ * sizeof(int32_t));
+
+	memset(hsof_dsm->buf.input, 0, DSM_FF_BUF_SZ * sizeof(int16_t));
+	memset(hsof_dsm->buf.output, 0, DSM_FF_BUF_SZ * sizeof(int16_t));
+	memset(hsof_dsm->buf.voltage, 0, DSM_FF_BUF_SZ * sizeof(int16_t));
+	memset(hsof_dsm->buf.current, 0, DSM_FF_BUF_SZ * sizeof(int16_t));
+
+	memset(hsof_dsm->buf.ff.buf, 0, DSM_FF_BUF_DB_SZ * sizeof(int32_t));
+	memset(hsof_dsm->buf.ff_out.buf, 0, DSM_FF_BUF_DB_SZ * sizeof(int32_t));
+	memset(hsof_dsm->buf.fb.buf, 0, DSM_FB_BUF_DB_SZ * sizeof(int32_t));
+
+	hsof_dsm->seq_ff = 0;
+	hsof_dsm->seq_fb = 0;
+	hsof_dsm->seq_sof_in = 0;
+
+	hsof_dsm->buf.ff.avail = DSM_FF_BUF_SZ;
+	hsof_dsm->buf.ff_out.avail = 0;
+	hsof_dsm->buf.fb.avail = 0;
+
+	return 0;
+}
+
+int sof_dsm_inf_create(struct sof_dsm_struct_t *hsof_dsm, struct comp_dev *dev)
+{
+	return sof_dsm_create(hsof_dsm, dev);
+}
+
+static int sof_dsm_inf_get_buffer(int32_t *buf, uint32_t frames,
+				  struct audio_stream *stream,
+				  int num_channel)
+{
+	int x, y;
+	union sof_dsm_buf input, output;
+
+	input.buf16 = (int16_t *)stream->r_ptr;
+	input.buf32 = (int32_t *)stream->r_ptr;
+	output.buf16 = (int16_t *)buf;
+	output.buf32 = (int32_t *)buf;
+
+	switch (stream->frame_fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		for (x = 0 ; x < frames ; x++) {
+			for (y = 0 ; y < num_channel ; y++) {
+				input.buf16 =
+					(int16_t *)_wrap_buf_ptr(input.buf16,
+								 stream);
+				output.buf16[num_channel * x + y] =
+					*input.buf16;
+				input.buf16++;
+			}
+		}
+		break;
+	case SOF_IPC_FRAME_S24_4LE:
+	case SOF_IPC_FRAME_S32_LE:
+		for (x = 0 ; x < frames ; x++) {
+			for (y = 0 ; y < num_channel ; y++) {
+				input.buf32 =
+					(int32_t *)_wrap_buf_ptr(input.buf32,
+								 stream);
+				output.buf32[num_channel * x + y] =
+					*input.buf32;
+				input.buf32++;
+			}
+		}
+		break;
+	default:
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static int sof_dsm_inf_put_buffer(int32_t *buf, uint32_t frames,
+				  struct audio_stream *stream, int num_channel,
+				  int num_slot, char mask, struct comp_dev *dev)
+{
+	int x, y, offset;
+	char value;
+	union sof_dsm_buf input, output;
+	const char pow_2[8] = {1, 2, 4, 8, 16, 32, 64, 128};
+
+	input.buf16 = (int16_t *)buf;
+	input.buf32 = (int32_t *)buf;
+	output.buf16 = (int16_t *)stream->w_ptr;
+	output.buf32 = (int32_t *)stream->w_ptr;
+
+	switch (stream->frame_fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		for (x = 0 ; x < frames ; x++) {
+			for (y = 0 ; y < num_slot ; y++) {
+				if (y == 0)
+					offset = 0;
+				output.buf16 =
+					(int16_t *)_wrap_buf_ptr(output.buf16,
+								 stream);
+				value = pow_2[y] & mask;
+				*output.buf16 =
+					value > 0 ? input.buf16[num_channel * x
+					+ (offset++)] : 0;
+				output.buf16++;
+			}
+		}
+		break;
+	case SOF_IPC_FRAME_S24_4LE:
+	case SOF_IPC_FRAME_S32_LE:
+		for (x = 0 ; x < frames ; x++) {
+			for (y = 0 ; y < num_slot ; y++) {
+				if (y == 0)
+					offset = 0;
```

Comment:
for (y = 0, offset = 0; y < num_slot; y++) ?

Reply:
Thanks. This condition check is no longer necessary because I refer chan_map information to check valid slot.